src/backends/dlib/dlib_actions.cpp

```diff
@@ -0,0 +1,135 @@
+
+        bool save_crops = false;
+        if (_params.has("save_crops")) {
+            save_crops = _params.get("save_crops").get<bool>();
+        }
+
+        // iterate image batch
+        for (size_t i = 0; i < vad.size(); i++) {
+            std::string uri = vad.at(i).get("uri").get<std::string>();
+
+            cv::Mat cvimg = imgs.at(i);
+            dlib::matrix <dlib::rgb_pixel> img;
+            dlib::assign_image(img, dlib::cv_image<dlib::rgb_pixel>(cvimg));
+
+            std::vector <APIData> ad_cls = vad.at(i).getv("classes");
+            std::vector <APIData> cad_cls;
+
+            // iterate bboxes per image
+            for (size_t j = 0; j < ad_cls.size(); j++) {
+                APIData bbox = ad_cls.at(j).getobj("bbox");
+                if (bbox.empty()) {
+                    throw ActionBadParamException("align/crop action cannot find bbox object for uri " + uri);
+                }
+                APIData ad_shape = bbox.getobj("shape");
+                if (ad_shape.empty()) {
+                    throw ActionBadParamException("align/crop action cannot find shape object for uri " + uri);
+                }
+
+                // adding bbox id
+                std::string bbox_id = genid(uri, "bbox" + std::to_string(j));
+                bbox_ids.push_back(bbox_id);
+                APIData ad_cid;
+                ad_cls.at(j).add(bbox_id, ad_cid);
+                cad_cls.push_back(ad_cls.at(j));
+                const dlib::rectangle shape_rect(ad_shape.get("left").get<long>(),
+                                                 ad_shape.get("top").get<long>(),
+                                                 ad_shape.get("right").get<long>(),
+                                                 ad_shape.get("bottom").get<long>());
+                std::vector <dlib::point> parts;
+                std::vector<double> points = ad_shape.get("points").get < std::vector < double >> ();
+                for (size_t idx = 0; idx < points.size() - 1; idx += 2) {
+                    parts.push_back(dlib::point(static_cast<long>(points[idx]), static_cast<long>(points[idx + 1])));
+                }
+
+                dlib::full_object_detection shape(shape_rect, parts);
+                dlib::matrix <dlib::rgb_pixel> r;
+                dlib::extract_image_chip(img, dlib::get_face_chip_details(shape, chip_size, bratio), r);
+                cv::Mat cropped_img = dlib::toMat(r);
+
+                // save crops if requested
+                if (save_crops) {
+                    std::string puri = dd_utils::split(uri, '/').back();
+                    cv::imwrite("crop_" + puri + "_" + std::to_string(j) + ".png", cropped_img);
+                }
+
+                // serialize crop into string (will be auto read by read_element in imginputconn)
+                std::vector<unsigned char> cropped_img_ser;
+                bool str_encoding = cv::imencode(".png", cropped_img, cropped_img_ser);
```

Comment:
you can now skip this encoding and directly pass the std::vector<cv::Mat> as a "data_raw_img" (line 128), see chain_actions.cc:100  . The image should be zero-copied, this should save space (of copy) and time (of encoding)

Reply:
Awesome, thanks!